File: ios/GuideStreamTV/Views/TeamCrestCircle.swift
//
//  TeamCrestCircle.swift
//  GuideStreamTV
//
//  Sports restyle (2026-09-25): a team crest on a filled circle with a white
//  ring. The fill is `sports_teams.fill_color` — the team colour, or its
//  secondary colour when the crest is mostly the primary one (computed
//  server-side by the sports_team_fill edge function) — falling back to the
//  ESPN primary colour on the game when the catalogue has no row yet.
//

import SwiftUI

enum SportsCrestStyle {
    /// White ring around every crest circle on the phone. Set from the
    /// approved mockup (3pt). Small circles scale it down — see `ring(for:)`.
    static let ringWidth: CGFloat = 3

    static func ring(for size: CGFloat) -> CGFloat {
        if size >= 44 { return ringWidth }
        if size >= 28 { return ringWidth * 0.67 }
        return max(1, ringWidth * 0.5)
    }

    /// Background of an unselected crest circle (picker tiles not followed).
    static let emptyFill = Color(hex: "141B26")
}

struct TeamCrestCircle: View {
    let team: GameTeam
    let size: CGFloat
    /// False draws the "not followed" state: dark fill and a faint 1.5pt ring.
    var filled: Bool = true
    var ringWidth: CGFloat? = nil

    private var fillHex: String? {
        SportsTeamCatalogService.shared.fillHex(forUid: team.uid) ?? team.primaryHex
    }

    private var fill: Color {
        guard filled else { return SportsCrestStyle.emptyFill }
        return fillHex.map { Color(hex: $0) } ?? Color.white.opacity(0.15)
    }

    var body: some View {
        Circle()
            .fill(fill)
            .overlay { crest.padding(size * 0.18) }
            .overlay {
                Circle().strokeBorder(
                    filled ? Color.white : Color.white.opacity(0.14),
                    lineWidth: filled ? (ringWidth ?? SportsCrestStyle.ring(for: size)) : 1.5
                )
            }
            .frame(width: size, height: size)
    }

    @ViewBuilder
    private var crest: some View {
        if let s = team.logoURL, !s.isEmpty, let url = URL(string: s) {
            AsyncImage(url: url) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .scaledToFit()
                        .shadow(color: .black.opacity(0.35), radius: 1.5, y: 1)
                case .failure:
                    abbreviation
                default:
                    Color.clear
                }
            }
        } else {
            abbreviation
        }
    }

    private var abbreviation: some View {
        Text(team.abbreviation)
            .font(.system(size: size * 0.26, weight: .heavy))
            .foregroundStyle(filled && Self.isLight(fillHex) ? Color(hex: "0B0F16") : .white)
            .lineLimit(1)
            .minimumScaleFactor(0.5)
    }

    /// True for fills light enough that white text would not read on them.
    static func isLight(_ hex: String?) -> Bool {
        guard var s = hex?.trimmingCharacters(in: .whitespaces) else { return false }
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6, let v = UInt64(s, radix: 16) else { return false }
        let r = Double((v >> 16) & 0xFF), g = Double((v >> 8) & 0xFF), b = Double(v & 0xFF)
        return (0.299 * r + 0.587 * g + 0.114 * b) > 160
    }
}

// MARK: - Shared Sports card parts (2026-09-25)

/// The one Watch button on every Sports card — hero, Live now and See all.
struct SportsWatchPill: View {
    var body: some View {
        Text("Watch ▶")
            .scaledFont(size: 11, weight: .bold)
            .foregroundStyle(.white)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(Capsule().fill(Color(hex: "F5821F")))
    }
}

/// Card background from `sports_games.image_url` — the ESPN game photo, or
/// the stadium photo the server falls back to — under a dark scrim so the
/// scores stay readable. No URL, or a failed load, keeps the plain card fill.
struct SportsPhotoBackdrop: View {
    let url: String?
    var cornerRadius: CGFloat = 16

    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: cornerRadius).fill(Color(hex: "161B27"))
            if let s = url, let u = URL(string: s) {
                AsyncImage(url: u) { phase in
                    if case .success(let image) = phase {
                        image.resizable().scaledToFill()
                            .overlay(
                                LinearGradient(
                                    colors: [Color(hex: "04090F").opacity(0.65), Color(hex: "04090F").opacity(0.88)],
                                    startPoint: .top, endPoint: .bottom
                                )
                            )
                    } else {
                        Color.clear
                    }
                }
                .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
            }
        }
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/sports/TeamCrestCircle.kt
package com.rork.guidestreamtvandroid.ui.sports

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil3.compose.SubcomposeAsyncImage
import com.rork.guidestreamtvandroid.data.models.SportsGame
import com.rork.guidestreamtvandroid.data.remote.SportsTeamCatalogService

/**
 * Sports restyle (2026-09-25): a team crest on a filled circle with a white
 * ring. The fill is `sports_teams.fill_color` — the team colour, or its
 * secondary colour when the crest is mostly the primary one (computed
 * server-side by sports_team_fill) — falling back to the game's ESPN primary
 * colour. Mirrors iOS TeamCrestCircle.swift.
 */
internal object SportsCrestStyle {
    /** White ring around every crest circle on the phone (approved mockup: 3). */
    val ringWidth: Dp = 3.dp

    fun ring(size: Dp): Dp = when {
        size >= 44.dp -> ringWidth
        size >= 28.dp -> ringWidth * 0.67f
        else -> maxOf(1.dp, ringWidth * 0.5f)
    }

    /** Background of a crest circle for a team that is not followed. */
    val emptyFill = Color(0xFF141B26)
}

@Composable
internal fun TeamCrestCircle(
    team: SportsGame.TeamSummary,
    size: Dp,
    modifier: Modifier = Modifier,
    filled: Boolean = true,
) {
    // Read the catalogue as state so circles recolour once it loads — the
    // callers' own parameters don't change when it does, so without this a
    // circle drawn before the catalogue arrived keeps the ESPN primary.
    val catalog = SportsTeamCatalogService.get()
    val catalogTeams by catalog.teams.collectAsState()
    val fillHex = remember(catalogTeams, team.uid, team.primaryHex) {
        catalog.fillHex(team.uid) ?: team.primaryHex
    }
    val fill = if (filled) hexToColor(fillHex, Color.White.copy(alpha = 0.15f)) else SportsCrestStyle.emptyFill
    val ring = if (filled) SportsCrestStyle.ring(size) else 1.5.dp
    val ringColor = if (filled) Color.White else Color.White.copy(alpha = 0.14f)
    val textColor = if (filled && fill.luminance() > 0.6f) Color(0xFF0B0F16) else Color.White
    val logoUrl = team.logoUrl?.takeIf { it.isNotBlank() }

    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            .background(fill)
            .border(ring, ringColor, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        val abbreviation: @Composable () -> Unit = {
            Text(
                team.abbreviation,
                fontSize = (size.value * 0.26f).sp,
                fontWeight = FontWeight.ExtraBold,
                color = textColor,
                maxLines = 1,
            )
        }
        if (logoUrl != null) {
            SubcomposeAsyncImage(
                model = logoUrl,
                contentDescription = team.abbreviation,
                modifier = Modifier.fillMaxSize().padding(size * 0.18f),
                contentScale = ContentScale.Fit,
                error = { Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { abbreviation() } },
                success = { state ->
                    Image(
                        painter = state.painter,
                        contentDescription = team.abbreviation,
                        modifier = Modifier.fillMaxSize(),
                        contentScale = ContentScale.Fit,
                    )
                },
            )
        } else {
            abbreviation()
        }
    }
}

// MARK: - Shared Sports card parts (2026-09-25)

/** The one Watch button on every Sports card — hero, Live now and See all. */
@Composable
internal fun SportsWatchPill() {
    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(com.rork.guidestreamtvandroid.ui.theme.BrandOrange)
            .padding(horizontal = 14.dp, vertical = 7.dp),
    ) {
        Text("Watch ▶", fontSize = 11.sp, fontWeight = FontWeight.Bold, color = Color.White)
    }
}

/**
 * Photo layer for a Sports card: `sports_games.image_url` (the ESPN game
 * photo, or the stadium photo the server falls back to) under a dark scrim so
 * the scores stay readable. Draws nothing without a URL or on a failed load,
 * leaving the card's own fill. Place first inside a Box, sized to the card.
 */
@Composable
internal fun SportsPhotoBackdrop(url: String?, modifier: Modifier = Modifier) {
    val u = url?.takeIf { it.isNotBlank() } ?: return
    SubcomposeAsyncImage(
        model = u,
        contentDescription = null,
        modifier = modifier,
        contentScale = ContentScale.Crop,
        success = { state ->
            Box(Modifier.fillMaxSize()) {
                Image(
                    painter = state.painter,
                    contentDescription = null,
                    modifier = Modifier.fillMaxSize(),
                    contentScale = ContentScale.Crop,
                )
                Box(
                    Modifier
                        .fillMaxSize()
                        .background(
                            androidx.compose.ui.graphics.Brush.verticalGradient(
                                0f to Color(0xFF04090F).copy(alpha = 0.65f),
                                1f to Color(0xFF04090F).copy(alpha = 0.88f),
                            ),
                        ),
                )
            }
        },
    )
}
